### DISCUSSION TITLE
NSLocalizedDescription The request is not supported

### DISCUSSION THREAD

**Discussion by Huy1908** [hbldh/bleak#1584] [Q&A]
I tried writing char but got the following error
"Failed to connect to iPhone: Failed to write characteristic 21: Error Domain=CBATTErrorDomain Code=6 "The request is not supported." UserInfo={NSLocalizedDescription=The request is not supported.}"

Please check the main file I have attached with the code content
[main.txt](https://github.com/hbldh/bleak/files/15504187/main.txt)

**Comment by dlech** [comment]
```python
import asyncio
from bleak import BleakScanner, BleakClient
import objc
from Foundation import NSBundle

async def scan_and_connect():
    devices = await BleakScanner.discover()

    print("Available devices:")
    for i, device in enumerate(devices):
        print(f"{i + 1}. {device.name} [{device.address}]")

    device_index = int(input("Select device to connect (enter index): ")) - 1
    selected_device = devices[device_index]

    async with BleakClient(selected_device.address) as client:
        try:
            if await client.is_connected():
```
An exception would be thrown on the `async with` line if we failed to connect, so `is_connected` is always true here. So no need to check it.

```python
                print(f"Connected to {selected_device.name} [{selected_device.address}]")

                services = await client.get_services()
                print("Services and Characteristics:")
                for service in services:
                    print(f"Service: {service.uuid}")
                    for char in service.characteristics:
                        print(f"  Characteristic: {char.uuid}, Properties: {char.properties}")

                char_uuid = "af0badb1-5b99-43cd-917a-a77bc549e3cc"

                await client.write_gatt_descriptor(char_uuid, b"\x41", response=True)
```
I think this should be `write_gatt_char` instead of `write_gatt_descriptor`.

```python
                value = await client.read_gatt_char(char_uuid)
                print("I/O Config Post-Write Value: {0}".format(value))
            else:
                print(f"Failed to connect to {selected_device.name}")
        except Exception as e:
            print(f"Failed to connect to {selected_device.name}: {e}")

async def main():
    await scan_and_connect()

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

Better to use `asyncio.run(main())` otherwise cleanup isn't run after a crash

**Comment by Huy1908** [comment]
I have also tried with write_gatt_char but the result is not better

**Reply by dlech** [reply]
Can you share the full output of the script (using the correct `write_gatt_char`)? And also include the stack trace of the exception so we can see what line is actually failing? Maybe the characteristic is not writeable? Or requires authentication or authorization?

**Reply by dlech** [reply]
Logging Bluetooth packets as described in the troubleshooting page of the docs should shed some light on the problem too.

**Reply by Huy1908** [reply]
**After I run the code the result is as follows:**

Available devices:
1. None [36B00EB0-E8E0-2B7F-4D83-759A72CE2C27]
2. iPhone [99E801D9-33FD-8CE8-4990-4B1DECD3C1D6]
3. None [08B2613F-0381-9299-C50A-DFFEECE56B03]
4. None [54E69FDA-337D-7BC5-F95C-8BE1144AB017]
Select device to connect (enter index): 2
/Users/Me/PY/send_data.py:103: FutureWarning: is_connected has been changed to a property. Calling it as an async method will be removed in a future version
  if await client.is_connected():
Connected to iPhone [99E801D9-33FD-8CE8-4990-4B1DECD3C1D6]
/Users/Me/PY/send_data.py:106: FutureWarning: This method will be removed future version, use the services property instead.
  services = await client.get_services()
Services and Characteristics:
Service: 0000180a-0000-1000-8000-00805f9b34fb
  Characteristic: 00002a29-0000-1000-8000-00805f9b34fb, Properties: ['read']
  Characteristic: 00002a24-0000-1000-8000-00805f9b34fb, Properties: ['read']
Service: d0611e78-bbb4-4591-a5f8-487910ae4366
  Characteristic: 8667556c-9a37-4c91-84ed-54ee27d90049, Properties: ['write', 'notify', 'extended-properties']
Service: 9fa480e0-4967-4542-9390-d343dc5d04ae
  Characteristic: af0badb1-5b99-43cd-917a-a77bc549e3cc, Properties: ['write', 'notify', 'extended-properties']
Service: 0000180f-0000-1000-8000-00805f9b34fb
  Characteristic: 00002a19-0000-1000-8000-00805f9b34fb, Properties: ['read', 'notify']
Service: 00001805-0000-1000-8000-00805f9b34fb
  Characteristic: 00002a2b-0000-1000-8000-00805f9b34fb, Properties: ['read', 'notify']
  Characteristic: 00002a0f-0000-1000-8000-00805f9b34fb, Properties: ['read']
Failed to connect to iPhone: Failed to write characteristic 21: Error Domain=CBATTErrorDomain Code=6 "The request is not supported." UserInfo={NSLocalizedDescription=The request is not supported.}

**Reply by dlech** [reply]
https://developer.apple.com/documentation/corebluetooth/cbatterror-swift.struct/code/requestnotsupported?language=objc

If you log Bluetooth packets, I expect that you will see that this error is coming from the iPhone in response to the write request, so that is where the problem is.

**Reply by Huy1908** [reply]
I want to send data from python code to iphone device, is there any solution?

**Reply by dlech** [reply]
Are you writing a custom program for the iPhone? If so, I would imagine it is just a matter of writing the iPhone program correctly to handle the request.
